tensorflow/java/src/main/java/org/tensorflow/Scope.java

```diff
@@ -0,0 +1,198 @@
+   * @return a new subscope
+   * @throws IllegalArgumentException if the name is invalid
+   */
+  public Scope withSubScope(String childScopeName) {
+    return toBuilder().nameScope(nameScope.withSubScope(childScopeName)).build();
+  }
+
+  /**
+   * Return a new scope that uses the provided name for an op.
+   *
+   * <p>Operations created within this scope will have a name of the form {@code
+   * name/opName[_suffix]}. This lets you name a specific operator more meaningfully.
+   *
+   * <p>Valid operator names must match the regular expression {@code [A-Za-z0-9.][A-Za-z0-9_.\\-]*}
+   *
+   * @param opName name for an operator in the returned scope
+   * @return a new Scope that uses opName for operations.
+   * @throws IllegalArgumentException if the name is invalid
+   */
+  public Scope withOpName(String opName) {
+    return toBuilder().nameScope(nameScope.withOpName(opName)).build();
+  }
+
+  /**
+   * Create a unique name for an operator, using a provided default if necessary.
+   *
+   * <p>This is normally called only by operator building classes.
+   *
+   * <p>This method generates a unique name, appropriate for the name scope controlled by this
+   * instance. Typical operator building code might look like
+   *
+   * <pre>{@code
+   * scope.graph().opBuilder("Constant", scope.makeOpName("Constant"))...
+   * }</pre>
+   *
+   * <p><b>Note:</b> if you provide a composite operator building class (i.e, a class that adds a
+   * set of related operations to the graph by calling other operator building code) you should also
+   * create a {@link #withSubScope(String)} scope for the underlying operators to group them under a
+   * meaningful name.
+   *
+   * <pre>{@code
+   * public static Stddev create(Scope scope, ...) {
+   *   // group sub-operations under a common name
+   *   Scope group = scope.withSubScope("stddev");
+   *   ... Sqrt.create(group, Mean.create(group, ...))
+   * }
+   * }</pre>
+   *
+   * @param defaultName name for the underlying operator.
+   * @return unique name for the operator.
+   * @throws IllegalArgumentException if the default name is invalid.
+   */
+  public String makeOpName(String defaultName) {
+    return nameScope.makeOpName(defaultName);
+  }
+
+  private Scope(Builder builder) {
+    graph = builder.graph;
+    if (builder.nameScope != null) {
+      nameScope = builder.nameScope;
+    } else {
+      nameScope = NameScope.create();
+    }
+  }
+
+  private final Graph graph;
+  private final NameScope nameScope;
+
+  private Builder toBuilder() {
+    return builder(graph).nameScope(nameScope);
+  }
+
+  private static Builder builder(Graph graph) {
+    return new Builder(graph);
+  }
+
+  private static final class Builder {
+    private Builder(Graph g) {
+      graph = g;
+    }
+
+    private Builder nameScope(NameScope ns) {
+      nameScope = ns;
+      return this;
+    }
+
+    private Scope build() {
+      return new Scope(this);
```

Comment:
Thank you for the thoughtful and detailed comments @asimshankar - much appreciated! I'll update this PR to reflect your suggestions.

May we sound you out a bit further on `Scope` and `NameScope`? No problem collapsing them in one class, though this what we were thinking.

There are two classes now, with `Scope` delegating to `NameScope`. We organized it as a main class delegating to a specialized class so we can add other self-contained groups of properties like "control dependencies" without putting everything in one implementation.

In this approach, the public `Scope` would be augmented to hold package private `NameScope` and `ControlDependency` objects, delegating to each as appropriate.

A `Scope` carrying a certain name hierarchy would then implement `withControlDependencies()` by reusing the existing `NameScope` and just creating a new `ControlDependency` object. `withName()` and `withOpName()` would do the opposite, etc. Having specialized classes to manage each group of properties would also make it a bit easier to test each in isolation.

What do you think?

[Re: Builders to construct classes rather than raw constructors; we were following [https://engdoc.corp.google.com/eng/doc/devguide/java/practices/builders.md?cl=head](https://engdoc.corp.google.com/eng/doc/devguide/java/practices/builders.md?cl=head) style that  prefers Builders or a static factory method to raw constructors. We chose Builders here as `NameScope` manages 3/4 constructor parameters of interest, and `Scope` currently has just one but we anticipate at least two after adding `ControlDependency`.  That said, we're happy to switch to raw constructors if you prefer this style for Tensorflow, we can switch 
 in this and subsequent PRs, please let us know.]